src/util/rocksdb/properties.rs

```diff
@@ -466,6 +479,251 @@ impl DecodeProperties for UserCollectedProperties {
+
+#[derive(Debug, Default)]
+pub struct RangeProperties {
+    pub total_size: u64,
+    pub total_keys: u64,
+    pub index_handles: IndexHandlesMap,
+}
+
+impl RangeProperties {
+    pub fn encode(&self) -> UserProperties {
+        let mut props = UserProperties::new();
+        props.encode_u64(PROP_TOTAL_SIZE, self.total_size);
+        props.encode_u64(PROP_TOTAL_KEYS, self.total_keys);
+        props.encode(PROP_RANGE_INDEX, self.index_handles.encode());
+        props
+    }
+
+    pub fn try_decode<T: DecodeProperties>(props: &T) -> Result<RangeProperties> {
+        match RangeProperties::decode(props) {
+            Ok(res)=>return Ok(res),
+            Err(e)=>warn!("old_version:can not decode to RangeProperties:{:?}, try to decode from SizeProperties",e),
+        }
+        SizeProperties::decode(props).map(|res| res.into())
+    }
+
+    pub fn decode<T: DecodeProperties>(props: &T) -> Result<RangeProperties> {
+        let mut res = RangeProperties::default();
+        res.total_size = props.decode_u64(PROP_TOTAL_SIZE)?;
+        res.total_keys = props.decode_u64(PROP_TOTAL_KEYS)?;
+        let v = props.decode(PROP_RANGE_INDEX)?;
+        res.index_handles = IndexHandlesMap::decode(v, 2)?;
+        Ok(res)
+    }
+
+    pub fn get_approximate_size_in_range(&self, start: &[u8], end: &[u8]) -> u64 {
+        self.index_handles
+            .get_approximate_distance_in_range(0, start, end)
+    }
+
+    pub fn get_approximate_keys_in_range(&self, start: &[u8], end: &[u8]) -> u64 {
+        self.index_handles
+            .get_approximate_distance_in_range(1, start, end)
+    }
+
+    pub fn smallest_key(&self) -> Option<Vec<u8>> {
+        self.index_handles.smallest_key()
+    }
+
+    pub fn largest_key(&self) -> Option<Vec<u8>> {
+        self.index_handles.largest_key()
+    }
+
+    pub fn merge(&mut self, other: RangeProperties) {
+        self.total_keys += other.total_keys;
+        self.total_size += other.total_size;
+        self.index_handles
+            .handles
+            .extend(other.index_handles.handles);
+    }
+
+    pub fn size_idx_iter(&self) -> IndexHandlesIterator {
+        self.index_handles.iter(0)
+    }
+}
+
+impl Into<RangeProperties> for SizeProperties {
+    fn into(self) -> RangeProperties {
+        RangeProperties {
+            total_size: self.total_size,
+            total_keys: 0,
+            index_handles: self.index_handles.into_index_handles_map(0, 2),
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct RangePropertiesCollector {
+    props: RangeProperties,
+    last_key: Vec<u8>,
+    size_index_handle: IndexHandle,
+    keys_index_handle: IndexHandle,
+}
+
+impl TablePropertiesCollector for RangePropertiesCollector {
+    fn add(&mut self, key: &[u8], value: &[u8], entry_type: DBEntryType, _: u64, _: u64) {
+        if entry_type != DBEntryType::Put {
+            return;
+        }
+
+        // keys
+        self.props.total_keys += 1;
+        self.keys_index_handle.size += 1;
```

Comment:
I think `size` is not a good name for keys, how about change `size` to `value` in the `IndexHandle`?